client/http.go

```diff
@@ -0,0 +1,156 @@
+	c := &httpClient{
+		root:   url,
+		group:  group,
+		client: client,
+	}
+	return c, nil
+}
+
+// httpClient implements Client through http requests to a Drand relay.
+type httpClient struct {
+	root   string
+	client HTTPClient
+	group  *key.Group
+}
+
+// FetchGroupInfo attempts to initialize an httpClient when
+// it does not know the full group paramters for a drand group.
+func (h *httpClient) FetchGroupInfo(groupHash []byte) (*key.Group, error) {
+	// fetch the `Group` to validate connectivity.
+	groupResp, err := h.client.Get(fmt.Sprintf("%s/group", h.root))
+	if err != nil {
+		return nil, err
+	}
+
+	protoGrp := drand.GroupPacket{}
+	if err := json.NewDecoder(groupResp.Body).Decode(&protoGrp); err != nil {
+		return nil, err
+	}
+	grp, err := key.GroupFromProto(&protoGrp)
+	if err != nil {
+		return nil, err
+	}
+
+	if grp.PublicKey == nil {
+		return nil, fmt.Errorf("Group does not have a valid key for validation")
+	}
+
+	if groupHash != nil && !bytes.Equal(grp.Hash(), groupHash) {
+		return nil, fmt.Errorf("%s does not advertise the expected drand group", h.root)
+	}
+	return grp, nil
+}
+
+// Implement textMarshaller
+func (h *httpClient) MarshalText() ([]byte, error) {
+	return json.Marshal(h)
+}
+
+type randData struct {
+	Round             uint64 `json:"round,omitempty"`
+	Signature         []byte `json:"signature,omitempty"`
+	PreviousSignature []byte `json:"previous_signature,omitempty"`
+	Randomness        []byte `json:"randomness,omitempty"`
+}
+
+// Get returns a the randomness at `round` or an error.
+func (h *httpClient) Get(ctx context.Context, round uint64) (Result, error) {
+	randResponse, err := h.client.Get(fmt.Sprintf("%s/public/%d", h.root, round))
+	if err != nil {
+		return Result{}, err
+	}
+
+	randResp := randData{}
+	if err := json.NewDecoder(randResponse.Body).Decode(&randResp); err != nil {
+		return Result{}, err
+	}
+	if len(randResp.Signature) == 0 || len(randResp.PreviousSignature) == 0 {
+		return Result{}, fmt.Errorf("insufficent response")
+	}
+
+	b := beacon.Beacon{
+		PreviousSig: randResp.PreviousSignature,
+		Round:       randResp.Round,
+		Signature:   randResp.Signature,
+	}
+	if err := beacon.VerifyBeacon(h.group.PublicKey.Key(), &b); err != nil {
+		return Result{}, err
+	}
+
+	return Result{Round: randResp.Round, Signature: randResp.Signature}, nil
+}
+
+// Watch returns new randomness as it becomes available.
+func (h *httpClient) Watch(ctx context.Context) <-chan Result {
+	ch := make(chan Result, 1)
+	r := h.RoundAt(time.Now())
+	val, err := h.Get(ctx, r)
+	if err != nil {
+		ch <- val
+	} else {
+		close(ch)
+		return ch
+	}
+	time.AfterFunc(time.Unix(beacon.TimeOfRound(h.group.Period, h.group.GenesisTime, r+1), 0).Sub(time.Now()), func() {
+		t := time.NewTicker(h.group.Period)
+		for {
+			select {
+			case <-t.C:
+				r, err := h.Get(ctx, h.RoundAt(time.Now()))
+				if err != nil {
+					ch <- r
+				}
+				// TODO: keep trying on errors?
```

Comment:
Log the error at minimum?

How about return an error channel in addition to the result channel? This seems to be the way go-ipfs is going https://github.com/ipfs/interface-go-ipfs-core/pull/62

Reply:
What is the consumer going to do in this case? especially when the client is a hybrid of http and pub/sub, i think there isn't an interface for errors that a consumer will be able to do anything actionable with. I'll log for now

Reply:
I’m not sure, give some feedback to the user that the randomness has been delayed? Give an option to quit? It just doesn’t feel right to not give the caller some way of handling the error even it’s likely to just be ignored...

Reply:
how would you structure this in the context of the pubsub watch? the interface will need to be a hybrid of both